### PR TITLE
issue/3163 Added correctToPass and _hasItemScoring support

### PR DESCRIPTION
**Pull request by oliverfoster**
https://github.com/adaptlearning/adapt_framework/issues/3163

### Added
* Added `correctToPass` to compliment `scoreToPass`

### Changed
* Use `QuestionModel.score`, `QuestionModel.maxScore` and `QuestionModel.minScore` rather than `_questionWeight` and `_isCorrect`

Requires https://github.com/adaptlearning/adapt_framework/pull/3172

### CODE REVIEW

**Review by cahirodoherty-learningpool — APPROVED** [review]
👀

File: js/adapt-assessmentArticleModel.js [hunk 1]
```diff
@@ -363,11 +372,14 @@ define([
...
       var scoreToPass = assessmentConfig._scoreToPass;
+      const correctToPass = assessmentConfig._correctToPass || 0;
 
       var scoreAsPercent = this.get('_scoreAsPercent');
       var score = this.get('_score');
+      const correctAsPercent = this.get('_correctAsPercent');
+      const correctCount = this.get('_correctCount');
 
-      var isPass = isPercentageBased ? (scoreAsPercent >= scoreToPass) : (score >= scoreToPass);
+      const isPass = isPercentageBased ? (scoreAsPercent >= scoreToPass && correctAsPercent >= correctToPass) : (score >= scoreToPass && correctCount >= correctToPass);
```
Comment (eleanor-heath) [inline js/adapt-assessmentArticleModel.js:382]: Can we format this ternary statement a bit better and separate it out onto different lines like how you have formatted this one

```
const isPass = !isComplete
? false
: (assessmentsConfig._isPercentageBased !== false)
  ? scoreAsPercent >= scoreToPass && correctAsPercent >= correctToPass
  : score >= scoreToPass && correctCount >= correctToPass;
```

**Review by eleanor-heath — APPROVED** [review]

File: properties.schema [hunk 2]
```diff
@@ -36,6 +36,15 @@
                   "default": 60,
                   "help": "The minimum score required by the learner to pass the assessment.",
                   "validators": ["number"]
+                },
+                "_correctToPass": {
+                  "type": "number",
+                  "required": false,
+                  "title": "Correct mark",
```
Comment (tomgreenfield) [inline properties.schema:43]:
```suggestion
                  "title": "Correct pass mark",
```

File: properties.schema [hunk 3]
```diff
@@ -36,6 +36,15 @@
...
+                  "inputType": "Number",
+                  "default": 60,
+                  "help": "The minimum number of correct questions required by the learner to pass the assessment.",
```
Comment (tomgreenfield) [inline properties.schema:46]: Should this mention that this is/could be a percentage?

File: properties.schema [hunk 4]
```diff
@@ -95,6 +104,15 @@
                   "help": "The minimum score required by the learner to pass the assessment.",
                   "validators": ["number"]
                 },
+                "_correctToPass": {
+                  "type": "number",
+                  "required": false,
+                  "title": "Correct mark",
```
Comment (tomgreenfield) [inline properties.schema:110]:
```suggestion
                  "title": "Correct pass mark",
```

File: js/assessment.js [hunk 5]
```diff
@@ -385,6 +392,9 @@ define([
         maxScore: maxScore,
         score: score,
         scoreToPass: assessmentsConfig._scoreToPass,
```
Comment (tomgreenfield) [inline js/assessment.js:394]: `scoreToPass` and `isPercentageBased` are already available in variables above.

Should `correctToPass` also be surfaced in the state?

File: js/adapt-assessmentArticleModel.js [hunk 6]
```diff
@@ -722,6 +746,22 @@ define([
       });
       this.set('_questions', questions);
```
Comment (tomgreenfield) [inline js/adapt-assessmentArticleModel.js:747]: Can be included in the `set()` starting on line 749.

File: README.md [hunk 7]
```diff
@@ -39,11 +39,13 @@ With the [Adapt CLI](https://github.com/adaptlearning/adapt-cli) installed, run
...
+>**\_scoreToPass** (integer): This is the achievement score required to pass the assessment. The learner's score must be greater than or equal to this score. It is the cumulative raw score needed to pass unless **\_isPercentageBased** is set to `true`.  
 
->**\_isPercentageBased** (boolean): Determines whether the value of **\_scoreToPass** should be treated as a percentage or as the raw score. For example, if **\_isPercentageBased** is set to `true`, a **\_scoreToPass** value of `60` will be treated as `60%`.
+>**\_correctToPass** (integer): This is the achievement correctness required to pass the assessment. The learner's score must be greater than or equal to this score. It is the cumulative raw correctness needed to pass unless **\_isPercentageBased** is set to `true`.  
+
+>**\_isPercentageBased** (boolean): Determines whether the values of **\_scoreToPass** and **\_correctToPass** should be treated as a percentage or as the raw score and correctness. For example, if **\_isPercentageBased** is set to `true`, a **\_scoreToPass** value of `60` will be treated as `60%`.
```
Comment (tomgreenfield) [inline README.md:48]:
```suggestion
>**\_isPercentageBased** (boolean): Determines whether the values of **\_scoreToPass** and **\_correctToPass** should be treated as percentages or as the raw score and correctness. For example, if **\_isPercentageBased** is set to `true`, a **\_scoreToPass** value of `60` will be treated as `60%`.  
```

File: README.md [hunk 8]
```diff
@@ -60,7 +62,9 @@ The following attributes are appended to a particular article within *articles.j
 
 >**\_scoreToPass** (number): This is the achievement score required to pass the assessment. The learner's score must be greater than or equal to this score. It is the cumulative raw score needed to pass unless **\_isPercentageBased** is set to `true`.    
 
->**\_isPercentageBased** (boolean): Determines whether the value of **\_scoreToPass** should be treated as a percentage or as the raw score. For example, if **\_isPercentageBased** is set to `true`, a **\_scoreToPass** value of `60` will be treated as `60%`.   
+>**\_correctToPass** (number): This is the achievement correctness required to pass the assessment. The learner's correctness must be greater than or equal to this value. It is the cumulative raw correctness needed to pass unless **\_isPercentageBased** is set to `true`.    
+
+>**\_isPercentageBased** (boolean): Determines whether the valuex of **\_scoreToPass** and **\_correctToPass** should be treated as a percentage or as the raw score and correctness. For example, if **\_isPercentageBased** is set to `true`, a **\_scoreToPass** value of `60` will be treated as `60%`.
```
Comment (tomgreenfield) [inline README.md:67]:
```suggestion
>**\_isPercentageBased** (boolean): Determines whether the values of **\_scoreToPass** and **\_correctToPass** should be treated as percentages or as the raw score and correctness. For example, if **\_isPercentageBased** is set to `true`, a **\_scoreToPass** value of `60` will be treated as `60%`.   
```

File: properties.schema [hunk 9]
```diff
@@ -83,7 +92,7 @@
                   "default": true,
                   "title": "Percentage based",
                   "inputType": "Checkbox",
-                  "help": "Determines whether the value of 'Pass mark' should be treated as a percentage (default) or as the raw score.",
+                  "help": "Determines whether the values of the 'Pass mark' should be treated as a percentage (default) or as the raw score and correctness.",
```
Comment (tomgreenfield) [inline properties.schema:95]:
```suggestion
                  "help": "Determines whether the pass mark values should be treated as percentages (default) or as the raw score and correctness.",
```

**Review by tomgreenfield — CHANGES_REQUESTED** [review]

**Review by tomgreenfield — APPROVED** [review]
:eye:

**Pull request merged**